gcc8 fixes

sm00th opened the pull request:
This is my attempt at fixing #865 and #866. I tried to keep the commits small so that it would be easier to follow and fix if some of the approaches are sub-optimal.

Build/load/unload/run tested with [unlikely-likely.patch.txt](https://github.com/dynup/kpatch/files/2097337/unlikely-likely.patch.txt) on top of 4.17 which covers both cases. Also build-tested with all of the integration-tests we have and it now builds all of the patches gcc7 does successfully.

Note: if the patch causes a function to change subsections (.e.g. .text -> .text.unlikely) it will be marked as CHANGED and included in the output even if the actual contents of the function were not changed. This makes sense to me but can be a source of confusion I guess.

sm00th commented:
> Do you expect many more parent/child types than just ".cold." ?

I am yet to meet it in the wild and failing to navigate gcc's sources properly but I suspect there might be a '.hot.' counterpart, or at least it is possible it will be added in the future.

jpoimboe commented:
> I am yet to meet it in the wild and failing to navigate gcc's sources properly but I suspect there might be a '.hot.' counterpart, or at least it is possible it will be added in the future.

/me cries.  As objtool maintainer I've never seen '.hot.' (and I'm pretty certain I would have heard about it already if it's out there).  Maybe that's something we can look forward to next year for GCC 9.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -608,15 +637,41 @@ static void kpatch_compare_sections(struct list_head *seclist)
+		return SUBSECTION_UNLIKELY;
+	} else if (!strncmp(sec->name, ".text.hot.", 10)) {
+		return SUBSECTION_HOT;
+	}
```
else's aren't necessary after returns

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -608,15 +637,41 @@ static void kpatch_compare_sections(struct list_head *seclist)
 	}
 }
 
+static sstype_t kpatch_section_type(struct section *sec)
```
Section type already has another meaning.  Maybe 'kpatch_subsection_type'?

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -608,15 +637,41 @@ static void kpatch_compare_sections(struct list_head *seclist)
+{
+	if (strncmp(sec->name, ".text.", 6)) {
+		DIFF_FATAL("unsupported section type: %s", sec->name);
+	}
```
I think this DIFF_FATAL could be confusing, based on how this function is called.  For example, if were a data section, this error would hit instead of the "symbol changed sections" error.  So maybe just remove this check altogether and assume that it's just always SUBSECTION_NORMAL if it doesn't match the other string comparisons?

jpoimboe commented:
Some unit tests would be good for this.

sm00th commented:
> - Does it handle the case where a child function appears or disappears as a result of the patch? That would mean the parent should be marked changed (and the child, if it exists, included in the patch).

It does, however all of the patches I tested with resulted in parent function being changed as well. It won't mark the parent changed if just the .cold. subfunction appears/disappears, but I am under an impression that it can't happen. Might this assumption be wrong?

> - If only the parent changes, but not the child, does the child still get pulled into the patch?

It does not. Won't the patched function be able to use the already loaded .cold. subfunction?

> - Some unit tests would be good for this.

Yes, I was planning to add them once all of the major missing parts are included in the pull-request.

jpoimboe commented:
> It does, however all of the patches I tested with resulted in parent function being changed as well. It won't mark the parent changed if just the .cold. subfunction appears/disappears, but I am under an impression that it can't happen. Might this assumption be wrong?

Yes, the subfunction can appear/disappear.  The subfunction is basically a subset of the function which the compiler deems unlikely to execute, so it moves part of the function to the unlikely text section, thus creating the "subfunction".  If the patch file removes that piece of unlikely code from the function, the subfunction will disappear.  Similarly, if the patch file adds a piece of unlikely to a function, it may be placed in a (new) subfunction.

> If only the parent changes, but not the child, does the child still get pulled into the patch?
> It does not. Won't the patched function be able to use the already loaded .cold. subfunction?

Yes, I suppose that would be fine.  Either way works.

sm00th commented:
> If the patch file removes that piece of unlikely code from the function, the subfunction will disappear. Similarly, if the patch file adds a piece of unlikely to a function, it may be placed in a (new) subfunction.

Yes, but the thing is every time I tried to just add a cold subfunction (if (!jiffies) printk) the parent function got changed as well. Do you happen to know a way to do this without affecting parent function so that we can add it to unittests?

jpoimboe commented:
> Yes, but the thing is every time I tried to just add a cold subfunction (if (!jiffies) printk) the parent function got changed as well. Do you happen to know a way to do this without affecting parent function so that we can add it to unittests?

Yeah, find a function which already has a subfunction and figure out which part of the function's source corresponds to the subfunction.  It will presumably be in an "unlikely" clause.  Then just add an asm("nop") to that clause.  That should in theory modify the subfunction without modifying the parent function, assuming there aren't any other unlikely clauses after the changed one (as that would affect some other relocations in the parent function).

sm00th commented:
> Yeah, find a function which already has a subfunction and figure out which part of the function's source corresponds to the subfunction. It will presumably be in an "unlikely" clause. Then just add an asm("nop") to that clause. That should in theory modify the subfunction without modifying the parent function, assuming there aren't any other unlikely clauses after the changed one (as that would affect some other relocations in the parent function).

But that will only change the subfunction, not add or remove it. Anyway I changed it so that should now cover add/delete cases even if parent function is not change. Shouldn't hurt even if it never triggers.

sm00th commented:
v2:
  - Suggested codestyle and function names changes
  - added children skips to kpatch_reate_patches_sections() and function_ptr_rela()
  - addition/deletion of subfunction will now mark parent as changed
  - kpatch_mark_ignored_functions_same() will now mark children of ignored functions same

jpoimboe commented:
Sorry, I neglected to read and I completely missed your point about adding/removing subfunctions.  Adding/removing a subfunction will always change the parent, because the parent will always jump to the subfunction.  So adding/removing the subfunction will add/remove a jump to the parent function.

sm00th commented:
Okay, we can drop the last patch (create-diff-object: make deleted subfunctions mark parents as changed ) then.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -608,16 +665,40 @@ static void kpatch_compare_sections(struct list_head *seclist)
+		if (is_rela_section(sec))
+			sym = sec->base->sym;
+		else
+			sym = sec->sym;
```
optional: could probably do `sym = is_rela_section(sec) ? sec->base->sym : sec->sym;` which I think would be more succinct and readable

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -608,16 +665,40 @@ static void kpatch_compare_sections(struct list_head *seclist)
+			sym->status = sec->status;
+			if (sym->parent) {
+				if (sec->status == NEW && sym->parent->sec->twin)
+					sym->parent->status = CHANGED;
```
Since we decided that a new subfunction will always change the parent, I think this isn't needed?

sm00th replied:
I think we better leave it here, otherwise we are dependent on the section ordering in the list. I.e. if the patch results in a NEW subfunction and parent section is already marked as CHANGED without this if we will be marking parent as NEW. Or am I missing something?

jpoimboe replied:
It seems a little convoluted, what do you think about changing it to something like this?
```
	/* sync symbol status */
	list_for_each_entry(sec, seclist, list) {
		if (is_rela_section(sec)) {
			if (sec->base->sym && sec->base->sym->status != CHANGED)
				sec->base->sym->status = sec->status;
		} else {
			struct symbol *sym = sec->sym;

			if (sym && sym->status != CHANGED)
				sym->status = sec->status;

			if (sym && sym->child && sym->status == SAME &&
			    sym->child->sec->status == CHANGED)
				sym->status = CHANGED;
		}
	}
```

sm00th replied:
That should work and is easier to read. Will use that, thanks.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -585,6 +622,26 @@ static int kpatch_line_macro_change_only(struct section *sec)
 }
 #endif
 
+static void kpatch_find_deleted_subsections(struct kpatch_elf *kelf)
```
I agree we can remove this

sm00th commented:
v3:
  - codestyle fixes
  - dropped "create-diff-object: make deleted subfunctions mark parents as changed" patch
  - kpatch_compare_sections() patch reworked to be less confusing
  - added unittests

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -612,12 +655,34 @@ static void kpatch_compare_sections(struct list_head *seclist)
+				sym->status = sec->status;
+
+			if (sym && sym->child && sym->status == SAME &&
+					sym->child->sec->status == CHANGED)
```
The alignment is off here, it should look like:
```
			if (sym && sym->child && sym->status == SAME &&
			    sym->child->sec->status == CHANGED)
```

sm00th replied:
Ah, my vim's tabstop/shiftwidth combination made it look like it was aligned.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -630,10 +695,12 @@ static void kpatch_compare_correlated_symbol(struct symbol *sym)
 	if (sym1->sec && sym2->sec && sym1->sec->twin != sym2->sec) {
-		if (sym2->sec->twin && sym2->sec->twin->ignore)
+		if ((sym2->sec->twin && sym2->sec->twin->ignore) ||
+				kpatch_subsection_changed(sym1->sec, sym2->sec))
```
another alignment issue here

sm00th commented:
v4:
  - indent fixes
